File: scripts/client_login.py
# coding=utf-8
import rospy
from login.srv import *


def client_login():
    # 初始化节点，指定名称
    rospy.init_node("client_login")
    # 等待指定的可用服务(阻塞)
    rospy.wait_for_service("login_info")

    # 输入信息
    name = raw_input("Input username:\n")
    passwd = raw_input("Input password:\n")

    try:
        # 构造客户端，指定服务端名称和传输的消息类型
        client = rospy.ServiceProxy("login_info", user)
        # 客户端发送消息，并返回结果
        resp = client.call(name, passwd)
        rospy.loginfo("%s", resp.state)
    except rospy.ServiceException, e:
        rospy.logwarn("Service call failed:%s" % e)


if __name__ == "__main__":
    client_login()

File: src/client_login.cpp
# include<ros/ros.h>
# include<login/user.h>


int main(int argc, char *argv[])
{
    // 初始化节点并指定节点名称
    ros::init(argc, argv, "client_login");
    // 创建句柄
    ros::NodeHandle nh;
    // 创建客户端并指定其发送的服务端名称
    ros::ServiceClient client = nh.serviceClient<login::user>("login_info");
    
    // 获取信息
    std::string name;
    std::string passwd;
    std::cout<<"Input username:"<<std::endl;
    std::cin>>name;
    std::cout<<"Input password:"<<std::endl;
    std::cin>>passwd;

    // 新建一个消息，传入内容
    login::user info;
    info.request.username = name;
    info.request.password = passwd;
    // 注意这里call函数返回的就是在服务端定义的回调函数的返回值(true or false)
    if (client.call(info))
    {
        ROS_INFO("%s",info.response.state.c_str());
    }else{
        ROS_ERROR("Failed to call service");
    }
    return 0;
}